pkg/controllers/monitoring-stack/components.go

```diff
@@ -0,0 +1,403 @@
+			},
+			{
+				APIGroups: []string{"extensions", "networking.k8s.io"},
+				Resources: []string{"ingresses"},
+				Verbs:     rbacVerbs,
+			},
+		},
+	}
+}
+
+func newServiceAccount(ms *stack.MonitoringStack) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ServiceAccount",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ms.Name + "-prometheus",
+			Namespace: ms.Namespace,
+		},
+	}
+}
+
+func newPrometheus(
+	ms *stack.MonitoringStack,
+	rbacResourceName string,
+	additionalScrapeConfigsSecretName string,
+	instanceSelectorKey string,
+	instanceSelectorValue string,
+) *monv1.Prometheus {
+	prometheusSelector := ms.Spec.ResourceSelector
+	if prometheusSelector == nil {
+		prometheusSelector = &metav1.LabelSelector{}
+	}
+	prometheus := &monv1.Prometheus{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Prometheus",
+			APIVersion: "monitoring.coreos.com/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ms.Name,
+			Namespace: ms.Namespace,
+			Labels:    prometheusLabels(ms.Name, instanceSelectorKey, instanceSelectorValue),
+		},
+
+		Spec: monv1.PrometheusSpec{
+			Image: nil,
+			// Prometheus does not use an Enum for LogLevel, so need to convert to string
+			LogLevel: string(ms.Spec.LogLevel),
+
+			Retention: ms.Spec.Retention,
+			Resources: ms.Spec.Resources,
+
+			ServiceAccountName: rbacResourceName,
+
+			ServiceMonitorSelector:          prometheusSelector,
+			ServiceMonitorNamespaceSelector: nil,
+			PodMonitorSelector:              prometheusSelector,
+			PodMonitorNamespaceSelector:     nil,
+
+			AdditionalScrapeConfigs: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: additionalScrapeConfigsSecretName,
+				},
+				Key: AdditionalScrapeConfigsSelfScrapeKey,
+			},
+		},
+	}
+	return prometheus
+}
+
+func newRoleBinding(ms *stack.MonitoringStack, rbacResourceName string) *rbacv1.RoleBinding {
+	roleBinding := &rbacv1.RoleBinding{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "rbac.authorization.k8s.io/v1",
+			Kind:       "RoleBinding",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rbacResourceName,
+			Namespace: ms.Namespace,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				APIGroup:  "",
+				Kind:      "ServiceAccount",
+				Name:      rbacResourceName,
+				Namespace: ms.Namespace,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     rbacResourceName,
+		},
+	}
+	return roleBinding
+}
+
+func newAdditionalScrapeConfigsSecret(name string, namespace string) *corev1.Secret {
+	return &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		StringData: map[string]string{
+			AdditionalScrapeConfigsSelfScrapeKey: `- job_name: prometheus-self
+  honor_labels: true
```

Comment:
I think it's better to define the self-scraping configuration like this so that we don't have to worry about the resource selectors defined for the monitoring stack. If we decide to go with a ServiceMonitor, the labels on that ServiceMonitor will need to be synthesised in a way to match what the user defines in the resourceSelector for the MonitorigStack.